### PR TITLE
Subsampling

### LINKED ISSUE #45: Add better sampling policies and custom GL object
Vectors can be sampled at a higher rate and further improve performance
Furthermore, there should be an option to display custom GL objects (cones, vectors etc) via graphics shaders

- [ ] custom shaders
- [x] better sampling (with object size adjustment)

### LINKED ISSUE #152: Precise dimension for cubic ergo cubes are not actually cubes
Cubic structures do not have correct dimensions - these should correspond to the ones specified in the file.
As a result cubes become more of a cuboid

### PR DESCRIPTION
**Pull request by LemurPwned**
Adding proper subsampling and changing fixing options
Fixes #152 
Resolves #45 
Now window is displayed if #133 or #125 occures.
This is not precisely solved now but definetely occurs in fewer cases.

### CODE REVIEW

File: Widgets/openGL_widgets/ArrowGLContext.py [hunk 1]
```diff
@@ -26,7 +27,7 @@ def __init__(self, data_dict, parent):
         self.ZERO_ROT = np.array([[1, 0, 0],
                                   [0, 1, 0],
                                   [0, 0, 1]])
-        self.SIDES = 32
+        self.SIDES = self.decimate
```
Comment (pawelkulig1) [inline Widgets/openGL_widgets/ArrowGLContext.py]: What reason is behind creating another variable? Can't we just use self.decimate instead?

File: Windows/PerfOptions.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def __init__(self, layer_size=None, object_type='None', parent=None):
             self.checkBox.setChecked(True)
 
         self.decimate = 1
-        self.averaging = 1
+        self.subsampling = 1
```
Comment (pawelkulig1) [inline Windows/PerfOptions.py]: Ok so averaging => subsampling is just refactor -> name change?

File: cython_modules/color_policy.pyx [hunk 3]
```diff
@@ -60,49 +59,10 @@ def hyper_contrast_calculation(np.ndarray[np.float32_t, ndim=3] color, xc, yc, z
...
-def construct_rotation_matrix(vector):
-    cos_x_rot = vector[1]
-    cos_y_rot = vector[0]/math.sqrt(1 - math.pow(vector[2],2))
-    sin_x_rot = math.sin(math.acos(cos_x_rot))  # radian input
-    sin_y_rot = math.sin(math.acos(cos_y_rot))
-    return np.array([[cos_y_rot, 0, sin_y_rot],
-                     [sin_y_rot*sin_x_rot, cos_x_rot, -sin_x_rot*cos_y_rot],
-                     [-cos_x_rot*sin_y_rot, sin_x_rot, sin_x_rot*cos_y_rot]])
-
 def process_vector_to_vbo(iteration,
```
Comment (pawelkulig1) [inline cython_modules/color_policy.pyx]: I understand that, you want to get rid of additional function and just paste code here. I think that this function should stay. It is usually better to leave more shorter functions than one blob. Maybe there is something more behind that?

File: cython_modules/cython_parse.pyx [hunk 4]
```diff
@@ -201,12 +201,35 @@ def decode_byte_size(byte_format_specification):
...
+            if yskip%yc == 0:           
+                if layer_skip is None:
+                    layer_skip = i     
+                zskip+=1
+                yskip = 0
+    if layer_skip is None:
+        layer_skip = i
+    return np.array(index_mask, dtype=np.int), layer_skip
+
+def cube(vec, spacer=0.1, dims=(0.1, 0.1, 0.1)):
```
Comment (pawelkulig1) [inline cython_modules/cython_parse.pyx]: Dims variable not used. Maybe you forgot to change spacer? Or is it just passed to have same arguments as cube2?

File: main.py [hunk 5]
```diff
@@ -363,10 +364,20 @@ def generalReceiver(self, options):
         self.doh.setDataObject(options, 'options')
         self.doh.setDataObject(self.screenshot_dir, 'screenshot_dir')
 
-        self.panes[self.current_pane].addWidget(\
-                self.sp.build_chain(self.current_widget_alias, self.doh, self))
+        try:
+            self.panes[self.current_pane].addWidget(\
+                    self.sp.build_chain(self.current_widget_alias, self.doh, self))
+        except (MemoryError, TimeoutError) as e:
+            x = PopUpWrapper("Insufficient resource", msg="You ran out of memory for this calculation"
```
Comment (pawelkulig1) [inline main.py]: Good one. Haven't thought about that.

File: cython_modules/color_policy.pyx (hunk 3 above)
Comment (LemurPwned) [reply cython_modules/color_policy.pyx]: no really, this function should be executed inline

File: Windows/PerfOptions.py (hunk 2 above)
Comment (LemurPwned) [reply Windows/PerfOptions.py]: Yup, averaging is deprecated 📦

File: Widgets/openGL_widgets/ArrowGLContext.py (hunk 1 above)
Comment (LemurPwned) [reply Widgets/openGL_widgets/ArrowGLContext.py]: this variable changes context depending on the class thats why it is assigned to class-specific variable

File: cython_modules/cython_parse.pyx (hunk 4 above)
Comment (LemurPwned) [reply cython_modules/cython_parse.pyx]: actually, spacer variable is not used. But good point, forgot to refactor so that spacer is removed

**Comment by LemurPwned** [comment]
Closing, needs more work on some bugs that occur

**Pull request closed without merging**